bin/runtime-common/src/messages.rs

```diff
@@ -340,8 +340,22 @@ pub mod source {
 		// if we're going to pay dispatch fee at the target chain, then we don't include weight
 		// of the message dispatch in the delivery transaction cost
 		let pay_dispatch_fee_at_target_chain = payload.dispatch_fee_payment == DispatchFeePayment::AtTargetChain;
+
+		let mut encoded_message_payload = payload.spec_version.encode();
+		encoded_message_payload.extend(match &payload.origin {
+			bp_message_dispatch::CallOrigin::TargetAccount(account_id, signer, signature) => {
+				let mut encoded_data = account_id.encode();
+				encoded_data.extend(signer.encode());
+				encoded_data.extend(signature.encode());
+				encoded_data
+			}
+			bp_message_dispatch::CallOrigin::SourceAccount(account_id) => account_id.encode(),
+			_ => Vec::<u8>::new(),
+		});
+		encoded_message_payload.extend(&payload.call);
+
 		let delivery_transaction = BridgedChain::<B>::estimate_delivery_transaction(
-			&payload.call,
+			&encoded_message_payload,
```

Comment:
I think we may use `payload.encode()` here instead of doing manual ^^^ encoding, right? ;)

Alternatively, there's other option which you may consider - instead of doing encoding (which also means extra allocations and memcpys), you may change implementation of `Size for MessagePayload` a bit [here](https://github.com/paritytech/parity-bridges-common/blob/master/primitives/message-dispatch/src/lib.rs#L133). So that it'll be including size of all other fields. This will also require `SourceChainAccountId`, `TargetChainAccountPublic`, `TargetChainSignature` to be `Size`d - you'll need to add implementation for basic types (`MultiSignature`,` `MultiSigner`, ...) somewhere [here](https://github.com/paritytech/parity-bridges-common/blob/master/primitives/runtime/src/lib.rs#L131).

This isn't required, though - I'll be happy with simple changing to `payload.encode()` :) Thanks!

Reply:
Thanks, that is a lot better. I changed the PR to just use `payload.encode()`

Reply:
Thanks!